gcc/config/riscv/rvp.md

```diff
@@ -17,11 +17,211 @@
 ;; along with GCC; see the file COPYING3.  If not see
 ;; <http://www.gnu.org/licenses/>.
 
-;; kabsw
+;; A list of the modes that are up to one word long vector.
+(define_mode_iterator VECI [(V4QI "!TARGET_64BIT") (V2HI "!TARGET_64BIT")
+                            (V8QI "TARGET_64BIT") (V4HI "TARGET_64BIT")
+			                      (V2SI "TARGET_64BIT")])
+(define_mode_iterator VQIHI [(V4QI "!TARGET_64BIT") (V2HI "!TARGET_64BIT")
+                             (V8QI "TARGET_64BIT") (V4HI "TARGET_64BIT")])
+(define_mode_iterator VSHI [(V2HI "!TARGET_64BIT") (V2SI "TARGET_64BIT")])
+(define_mode_attr VNHALF [(V2SI "SI") (V2HI "HI")])
+(define_mode_attr VSH_EXT [(V2SI "DI") (V2HI "HI")])
+
+;; <uk> expands to (un)signed (saturating) arithmetic operations
+(define_code_attr uk
+  [(plus "") (ss_plus "k") (us_plus "uk")
+   (minus "") (ss_minus "k") (us_minus "uk")])
+
+;; <bits> for specific bit number in 'simd' type instruction
+(define_mode_attr bits [(V8QI "8") (V4QI "8") (QI "8") (V4HI "16") (V2HI "16")
+			            (HI "16") (V2SI "32") (DI "64")])
+
+;; <all_plus> and <all_minus> for all types of addtion and subtraction
+(define_code_iterator all_plus [plus ss_plus us_plus])
+(define_code_iterator all_minus [minus ss_minus us_minus])
+
+;; k|(uk)|? add
+;; zpn is a mandatory subset for p extension, thus add32 (zprv subset) can also be matched
+(define_insn "<uk>add<mode>3"
+  [(set (match_operand:VECI 0 "register_operand"                "=r")
+	(all_plus:VECI (match_operand:VECI 1 "register_operand" " r")
+		       (match_operand:VECI 2 "register_operand" " r")))]
+  "TARGET_ZPN"
+  "<uk>add<bits>\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "<MODE>")])
+
+;; 64-bit
+(define_insn "rvp_<uk>adddi3"
+  [(set (match_operand:DI 0 "register_operand"              "=r")
+	(all_plus:DI (match_operand:DI 1 "register_operand" " r")
+		     (match_operand:DI 2 "register_operand" " r")))]
+  "TARGET_ZPSF"
+  "<uk>add64 %0, %1, %2"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; k|(uk)|? sub
+(define_insn "<uk>sub<mode>3"
+  [(set (match_operand:VECI 0 "register_operand"                  "=r")
+	(all_minus:VECI (match_operand:VECI 1 "register_operand" " r")
+			(match_operand:VECI 2 "register_operand" " r")))]
+  "TARGET_ZPN"
+  "<uk>sub<bits> %0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "<MODE>")])
+
+;; 64-bit
+(define_insn "rvp_<uk>subdi3"
+  [(set (match_operand:DI 0 "register_operand"               "=r")
+	(all_minus:DI (match_operand:DI 1 "register_operand" " r")
+		      (match_operand:DI 2 "register_operand" " r")))]
+  "TARGET_ZPSF"
+  "<uk>sub64 %0, %1, %2"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; uk|k add|sub w|h
```

Comment:
Could you simplify those pattern by `define_int_iterator` and `define_int_attr`?